packages/aws-cdk-lib/aws-cloudfront/lib/key-value-store.ts

```diff
@@ -0,0 +1,210 @@
+    };
+  }
+}
+
+/**
+ * An import source from a local file.
+ */
+export class AssetImportSource extends ImportSource {
+  private asset?: s3_assets.Asset;
+
+  /**
+   * @param path the path to the local file
+   * @param options the configuration for the temporarily created S3 file
+   */
+  constructor(public readonly path: string, private readonly options: s3_assets.AssetOptions = {}) {
+    super();
+  }
+
+  /**
+   * @internal
+   */
+  public _bind(scope: Construct): CfnKeyValueStore.ImportSourceProperty {
+    if (!this.asset) {
+      this.asset = new s3_assets.Asset(scope, 'ImportSource', {
+        path: this.path,
+        deployTime: true,
+        ...this.options,
+      });
+    } else if (Stack.of(this.asset) !== Stack.of(scope)) {
+      throw new Error(
+        `Asset is already associated with another stack '${Stack.of(this.asset).stackName}. ` +
+          'Create a new ImportSource instance for every stack.',
+      );
+    }
+
+    return {
+      sourceType: 'S3',
+      sourceArn: this.asset.bucket.arnForObjects(this.asset.s3ObjectKey),
+    };
+  }
+}
+
+/**
+ * The properties to create a Key Value Store.
+ */
+export interface KeyValueStoreProps {
+  /**
+   * The unique name of the Key Value Store.
+   *
+   * @default A generated name
+   */
+  readonly keyValueStoreName?: string;
+
+  /**
+   * A comment for the Key Value Store
+   *
+   * @default No comment will be specified
+   */
+  readonly comment?: string;
+
+  /**
+   * The import source for the Key Value Store.
+   *
+   * This will populate the initial items in the Key Value Store. The
+   * source data must be in a valid JSON format.
+   *
+   * @default No data will be imported to the store
+   */
+  readonly source?: ImportSource;
+}
+
+/**
+ * A CloudFront Key Value Store.
+ */
+export interface IKeyValueStore extends IResource {
+  /**
+   * The ARN of the Key Value Store.
+   *
+   * @attribute
+   */
+  readonly keyValueStoreArn: string;
+
+  /**
+   * The Unique ID of the Key Value Store.
+   *
+   * @attribute
+   */
+  readonly keyValueStoreId: string;
+
+  /**
+   * The status of the Key Value Store.
+   *
+   * @attribute
+   */
+  readonly keyValueStoreStatus: string;
+}
+
+/**
+ * A CloudFront Key Value Store.
+ *
+ * @resource AWS::CloudFront::KeyValueStore
+ */
+export class KeyValueStore extends Resource implements IKeyValueStore {
+
+  /**
+   * Import a Key Value Store using its ARN.
+   */
+  public static fromKeyValueStoreArn(scope: Construct, id: string, keyValueStoreArn: string): IKeyValueStore {
+    const storeId = Arn.split(keyValueStoreArn, ArnFormat.SLASH_RESOURCE_NAME).resourceName;
+    if (!storeId) {
+      throw new Error(`Invalid Key Value Store ID '${keyValueStoreArn}'`);
+    }
+    return new class Import extends Resource implements IKeyValueStore {
+      readonly keyValueStoreArn: string = keyValueStoreArn;
+      readonly keyValueStoreId: string = storeId!;
```

Comment:
Ugh i feel like TS should know that `storeId` exists here. Does it not?

Reply:
That's dumb. It does not. Oh well

Reply:
Yeah; it looks like TS realizes the type is `string` before the class definition but within it, it sees it as `string | undefined`